Upgrade `CORE_MSRV` to 1.82 and take advantage of some new features

ErichDonGubler opened the pull request:
**Connections**

With patches for [bug 1945020](https://bugzilla.mozilla.org/show_bug.cgi?id=1945020) landing against `mozilla-central`, we should be ready to bump `CORE_MSRV` to Rust 1.82. Woot! 🙌🏻

**Description**

Title says all. 💪

**Testing**

CI and done.

<!--
Thanks for filing! The codeowners file will automatically request reviews from the appropriate teams.

After you get a review and have addressed any comments, please explicitly re-request a review from the
person(s) who reviewed your changes. This will make sure it gets re-added to their review queue - you're no bothering us!
-->

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown`
- [x] Run `cargo xtask test` to run tests.
- [x] ~~Add change to `CHANGELOG.md`. See simple instructions inside file.~~

ErichDonGubler committed:
chore: remove `std::mem::*` imports now unnecessary with `CORE_MSRV`

`std::mem::{size,align}_of{,_val}` was added to `std::prelude` in Rust
1.80; see
[`rust`#123168](https://github.com/rust-lang/rust/pull/123168/).

ErichDonGubler committed:
build: update `CORE_MSRV` 1.76.0 → 1.80.1

ErichDonGubler committed:
refactor(hal): s/once_cell::Lazy/std::sync::LazyLock

Weaken our dependence on the `once_cell` crate by using functionality
from `std` instead that was upstreamed from `once_cell`, this time with
what's available in Rust 1.80+.

It's not yet possible to eliminate this dependency entirely, but do what
we can for now.

ErichDonGubler committed:
build: update `CORE_MSRV` 1.80.1 → 1.82.0

ErichDonGubler committed:
chore: satisfy `clippy::manual_c_str_literals`

This was previously an `allow`-by-default warning in Clippy's `pedantic`
group, but with Rust 1.83 it was promoted to a `warn`-by-default member
of its `complexity` group.

Co-Authored-By: Kevin Reid <kpreid@switchb.org>

ErichDonGubler committed:
refactor: use more `c` string literals

cwfitzgerald commented on naga/Cargo.toml:
```diff
@@ -15,7 +15,7 @@ autotests = false
 # them to have less restrictive MSRVs individually than the workspace as a
 # whole, if their code permits. See `../README.md` for details.
-rust-version = "1.76"
+rust-version = "1.80.0"
```
Typo? Shouldn't this be 1.82?

ErichDonGubler replied:
Ah, almost certainly. 👍🏻 Will fix.

ErichDonGubler replied:
I have several commits in this PR, and I'd correctly adjust `rust-version` for the 1.76 → 1.80 bump, but not the subsequent bump to 1.82. 😅 Fixed now, thanks!

ErichDonGubler committed:
refactor(naga): invert `coordinate` operand typeck to `is_none_or`

cwfitzgerald reviewed (changes requested):
Looks good, but I think you're missing a few things - I would grep for 1.76 and see what comes up, as hal's toml, the Readme, and probably a few other places need to be updated.

ErichDonGubler commented:
> Looks good, but I think you're missing a few things - I would grep for 1.76 and see what comes up, as hal's toml, the Readme, and probably a few other places need to be updated.

👍🏻 AFAIK, I've taken care of all of these as of my latest push. The `README` points to CI configuration as a reference, so I don't think we need to worry about that particular case again.

ErichDonGubler commented:
Gonna demote to a draft, while I figure out how to get crap compiling.

ErichDonGubler committed:
chore: satisfy `clippy::unnecessary_map_or` in Rust 1.85

ErichDonGubler committed:
chore(hal,core): use `core::{error::Error,format}` instead of `std`

This allows `wgpu-hal` to be used in `no_std` environments, except that
currently, only the `noop` backend supports `no_std`.
In the future, `cfg(all(gles, webgl))` should also be `no_std`, but that
requires further work.

Co-Authored_by: Kevin Reid <kpreid@switchb.org>

ErichDonGubler commented:
I've cut out any attempt to use `clippy::allow_attribute`, for the sake of letting this merge. That and updating to accommodate the latest Clippy configuration (viz., preferring `core` over `std`) should mean we're good to go, now!

cwfitzgerald commented on wgpu-core/Cargo.toml:
```diff
@@ -29,9 +29,6 @@ targets = [
 
-[lints.rust]
-unexpected_cfgs = { level = "warn", check-cfg = ['cfg(wgpu_validate_locks)'] }
-
```
I actually think this is the wrong way to move this change - we should be configuring lints in ideally one single place, instead of multiple. Could we drop this and do it separately.

ErichDonGubler replied:
Punted out of this PR with the latest push.

cwfitzgerald commented on wgpu-hal/src/gles/egl.rs:
```diff
@@ -474,7 +480,8 @@ struct Inner {
 // state of all our `EglContext`s. This forces us to track the number of such context to prevent
 // terminating the display if it's currently used by another `EglContext`.
-static DISPLAYS_REFERENCE_COUNT: Lazy<Mutex<HashMap<usize, usize>>> = Lazy::new(Default::default);
+static DISPLAYS_REFERENCE_COUNT: LazyLock<Mutex<HashMap<usize, usize>>> =
```
We should drop `once_cell` from the dependency list in `gles`

ErichDonGubler replied:
Ah, good catch! Implemented with latest push.

cwfitzgerald reviewed (approved):
Sorry wrong way
